File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/proc/AttOrElseProcessor.java
package ru.citeck.ecos.records3.record.op.atts.proc;

import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.commons.data.DataValue;
import ru.citeck.ecos.commons.data.ObjectData;

import java.util.*;

public class AttOrElseProcessor extends AbstractAttProcessor<List<DataValue>> {

    public static final String ATT_PREFIX = "a:";

    public AttOrElseProcessor() {
        super(true);
    }

    @Override
    protected Object processOne(@NotNull ObjectData meta,
                                @NotNull DataValue value,
                                @NotNull List<DataValue> arguments) {

        if (arguments.isEmpty() || value.isNotNull()) {
            return value;
        }

        for (DataValue orElseAtt : arguments) {

            if (orElseAtt.isTextual()) {
                String txtAtt = orElseAtt.asText();
                if (txtAtt.startsWith(ATT_PREFIX)) {
                    value = meta.get(txtAtt.substring(ATT_PREFIX.length()));
                } else {
                    value = orElseAtt;
                }
            } else {
                value = orElseAtt;
            }
            if (value.isNotNull() && (!value.isTextual() || !value.asText().isEmpty())) {
                return value;
            }
        }

        return value;
    }

    @Override
    protected List<DataValue> parseArgs(@NotNull List<DataValue> arguments) {
        return arguments;
    }

    @NotNull
    @Override
    public String getType() {
        return "or";
    }

    @NotNull
    @Override
    public Collection<String> getAttributesToLoad(@NotNull List<DataValue> arguments) {

        Set<String> attsToLoad = new HashSet<>();

        for (DataValue orElseAtt : arguments) {
            if (!orElseAtt.isTextual()) {
                continue;
            }
            String txtAtt = orElseAtt.asText();
            if (txtAtt.startsWith(ATT_PREFIX) && txtAtt.length() > ATT_PREFIX.length()) {
                attsToLoad.add(txtAtt.substring(ATT_PREFIX.length()));
            }
        }

        return attsToLoad;
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/atts/proc/AttOrElseProcessor.kt
package ru.citeck.ecos.records3.record.atts.proc

import ru.citeck.ecos.commons.data.DataValue
import ru.citeck.ecos.commons.data.ObjectData
import java.util.*

class AttOrElseProcessor : AttProcessor {

    companion object {
        const val ATT_PREFIX: String = "a:"
    }

    override fun process(attributes: ObjectData, value: DataValue, args: List<DataValue>): Any? {

        if (args.isEmpty()) {
            return value
        }
        if (value.isNotNull()) {
            if (!value.isTextual() || value.asText().isNotEmpty()) {
                return value
            }
        }

        var newValue = value
        for (orElseAtt in args) {
            newValue = if (orElseAtt.isTextual()) {
                val txtAtt: String = orElseAtt.asText()
                if (txtAtt.startsWith(ATT_PREFIX)) {
                    attributes.get(txtAtt.substring(ATT_PREFIX.length))
                } else {
                    orElseAtt
                }
            } else {
                orElseAtt
            }
            if (value.isNotNull() && (!value.isTextual() || value.asText().isNotEmpty())) {
                return value
            }
        }
        return newValue
    }

    override fun getAttsToLoad(arguments: List<DataValue>): Collection<String> {
        val attsToLoad: MutableSet<String> = HashSet()
        for (orElseAtt in arguments) {
            if (!orElseAtt.isTextual()) {
                continue
            }
            val txtAtt: String = orElseAtt.asText()
            if (txtAtt.startsWith(ATT_PREFIX) && txtAtt.length > ATT_PREFIX.length) {
                attsToLoad.add(txtAtt.substring(ATT_PREFIX.length))
            }
        }
        return attsToLoad
    }

    override fun getType(): String = "or"
}
